File: programs/tea/riesenie1.py
#zadanie 1: krizovaty
#autor: Roman Hudec
#datum: 6.3. 2022


from queue import PriorityQueue

class City:

    def __init__(self):

        self.graph = {}
        self.D = {}

    def add_house(self, house):

        if house not in self.graph:
            self.graph[house] = {}

    def add_route(self, house1, house2, distance, is_safe=True):

        self.add_house(house1)
        self.add_house(house2)

        if house2 not in self.graph[house1]:

            self.graph[house1][house2] = (distance, is_safe)
            self.graph[house2][house1] = (distance, is_safe)
            
        elif is_safe and not self.graph[house1][house2][1]:
            self.graph[house1][house2] = (distance, is_safe)
            self.graph[house2][house1] = (distance, is_safe)
            
        elif distance < self.graph[house1][house2][0]:
            self.graph[house1][house2] = (distance, is_safe)
            self.graph[house2][house1] = (distance, is_safe)

    def __contains__(self, house):
       return house in self.graph

    def __iter__(self):
        yield from self.graph

    def neighbours(self, house):

        yield from self.graph[house]

    def dijkstra(self, start):


        visited = set()
        pq = PriorityQueue()

        #dangerous_distance, distance, vertex -> put in the priority queue

        pq.put((0, 0, start))
        D = {}

        while not pq.empty():
            #print(D)

            dangerous_distance, distance, vertex = pq.get()

            if vertex not in visited:
                visited.add(vertex)
                D[vertex] = (dangerous_distance, distance)
                

                for n in self.neighbours(vertex):

                    if n not in visited:
                        weight, is_safe = self.graph[vertex][n]
                        #print(weight, is_safe)
                        
                        if is_safe:
                            pq.put((dangerous_distance, distance+weight, n))
                        else:
                            pq.put((dangerous_distance+weight, distance+weight, n))
        #print(D)
        self.D[start] = D                    
        return D

    def get_path(self, start, end):

        D = self.D.get(start, self.dijkstra(start))

        if end not in D:
            return -1, -1

        dangerous, distance = D[end]

        return dangerous, distance


if __name__ == "__main__":

    krizovatky, cesty, otazky = map(int, input().split())
    city = City()

    for i in range(cesty):

        odkial, kam, hodnota, bezpecie = input().split()

        bezpecie = (bezpecie == "B")

        city.add_route(int(odkial), int(kam), int(hodnota), bezpecie)

    for j in range(otazky):
        odkial, kam = input().split()
        
        print(*city.get_path(int(odkial), int(kam)))
        

File: programs/tea/riesenie1.cpp
//zadanie 1: krizovatky
//autor: Roman Hudec
//datum: 10.3.2022

#include <iostream>
#include <tuple>
#include <unordered_map>
#include <queue>
#define triple tuple<long long, long long, long long>

using namespace std;

struct compare{

    bool operator()(const triple &a, const triple &b){
        return a > b;
    }
};



class City{
    unordered_map<long long,  unordered_map<long long, pair<long long, bool> > > graph;

public:
    void addHouse(long long house){
        //ak tam este nie je, pridaj ho aj s jeho hash mapou
        if(graph.find(house) == graph.end()){
            graph[house] = unordered_map<long long, pair<long long, bool> >();
        }
    }

    void addRoute(long long house1, long long house2, long long distance, bool isSafe){
        addHouse(house1);
        addHouse(house2);

        //ak este nikdy medzi nimi hrana nebola
        if(graph[house1].find(house2) == graph[house1].end()){
            graph[house1][house2] = pair<long long, bool> (distance, isSafe);
            graph[house2][house1] = pair<long long, bool> (distance, isSafe);

        }else{
            pair<long long, bool> edge = graph[house1][house2];

            //nahradzam hranu, ak mame rovnaku boolean hodnotu alebo nova je okay, ale stara nebola
            if((isSafe == edge.second && edge.first > distance) || (isSafe && !edge.second)){
                graph[house1][house2] = pair<long long, bool> (distance, isSafe);
                graph[house2][house1] = pair<long long, bool> (distance, isSafe);
            }

        }
    }

    unordered_map<long long, pair<long long, long long>> dijkstra(long long start){

        //long long N = graph.size(); //pocet vrcholov
        unordered_map<long long, pair<long long, long long>> D; //pole vzdialenosti


        //long long visited[N];
        unordered_map<long long, bool> visited;

        for (auto vertex : graph){
            visited[vertex.first] = false;
        }

        /*for(long long i=0; i < N; i++){
            visited[i] = -1;
        }*/

        priority_queue<triple, vector<triple>, compare> pq;

        pq.push(triple(0, 0, start)); //pridavame trojicu (nebezpecie, bezpecie, vrchol)

        while(!pq.empty()){
            triple first = pq.top();
            pq.pop();

            long long dangerousDistance = get<0>(first);
            long long distance = get<1>(first);
            long long v = get<2>(first);

            if(visited[v] == false){
                D[v] = pair<long long, long long> (dangerousDistance, distance);
                visited[v] = true;

                for(auto n : graph[v]){ //prvky z hash mapy
                    long long w = n.first; //sused v-cka
                    if(visited[w] == false){
                        if(n.second.second){ // safe/not safe
                            long long weight = n.second.first;
                            pq.push(triple(dangerousDistance, distance+weight, w));
                        }else{
                            long long weight = n.second.first;
                            pq.push(triple(dangerousDistance+weight, distance+weight, w));
                        }


                    }
                }
            }
        }
        return D;
    }

    pair<long long, long long> getPath(long long start,long long end){

        if(graph.find(start) == graph.end()) return pair<long long, long long> (-1, -1);

        unordered_map<long long, pair<long long, long long>> D = dijkstra(start);


        if(D.find(end) == D.end()){
            return pair<long long, long long> (-1, -1);
        }

        //pair<long long, long long> output = D[end];

        return D[end];
    }
};

int main() {


    long long vrcholy, hrany, otazky;
    cin >> vrcholy >> hrany >> otazky;
    City city;

    for(long long m = 0; m < hrany; m++){
        long long zac, kon, vaha;
        bool ok;
        char pis;

        cin >> zac >> kon >> vaha >> pis;
        ok = (pis == 'B');

        city.addRoute(zac, kon, vaha, ok);
    }

    for(long long q = 0; q < otazky; q++){
        long long zac, kon;
        cin >> zac >> kon;
        pair<long long, long long> odpoved = city.getPath(zac, kon);
        cout << odpoved.first << " " <<odpoved.second << endl;

    }
    return 0;
}
